nova/scheduler/mixins.py

```diff
@@ -0,0 +1,47 @@
+# Copyright (c) 2021 SAP SE
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+import time
+
+from nova import context
+from nova.scheduler import client
+
+
+class HypervisorSizeFilterMixin(object):
+
+    _HV_SIZE_CACHE_RETENTION_TIME = 10 * 60
+
+    def __init__(self):
+        self._HV_SIZE_CACHE = {}
+
+    def _get_hv_size(self, host_state):
+        # expire the cache 10min after last write
+        time_diff = time.time() - self._HV_SIZE_CACHE.get('last_modified', 0)
+        if time_diff > self._HV_SIZE_CACHE_RETENTION_TIME:
+            self._HV_SIZE_CACHE = {}
+
+        if host_state.uuid not in self._HV_SIZE_CACHE:
+            scheduler_client = client.SchedulerClient()
+            placement_client = scheduler_client.reportclient
+            elevated = context.get_admin_context()
+            res = placement_client._get_inventory(elevated, host_state.uuid)
```

Comment:
Despite what I said in the [previous comment](https://github.com/sapcc/nova/pull/229/files#r695522892), I wonder if it wouldn't make sense to do that in batch, or possibly the opposite, spreading those requests a bit.
I am concerned, that we will get a stampede of a lot of small api calls.

As far as I can see, there is no batch api call, but we could ensure that the requests are a bit spread out by adding some random time to the expiry for each item.

Reply:
If we do single-item expiry, that would make sense. But yes, placement does not allow us to fetch inventory of multiple resource-providers at once. The API is built to get stampedes basically ;)